src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -720,8 +697,84 @@ public ParameterLookupValue(JsonPropertyInfo jsonPropertyInfo)
             public JsonPropertyInfo JsonPropertyInfo { get; }
         }
 
+        internal void InitializePropertyCache()
+        {
+            Debug.Assert(Kind == JsonTypeInfoKind.Object);
+
+            if (_properties != null)
+            {
+                // Properties have been exported to a metadata resolver,
+                // invalidate the property cache and build from scratch
+
+                ExtensionDataProperty = null;
+                if (PropertyCache is null)
+                {
+                    PropertyCache = CreatePropertyCache(capacity: _properties.Count);
+                }
+                else
+                {
+                    PropertyCache.Clear();
+                }
+
+                bool isOrderSpecified = false;
+                foreach (JsonPropertyInfo property in _properties)
+                {
+                    if (property.IsExtensionData)
+                    {
+                        if (ExtensionDataProperty != null)
+                        {
+                            ThrowHelper.ThrowInvalidOperationException_SerializationDuplicateTypeAttribute(Type, typeof(JsonExtensionDataAttribute));
+                        }
+
+                        ExtensionDataProperty = property;
+                        continue;
+                    }
+
+                    if (!PropertyCache.TryAddValue(property.Name, property))
+                    {
+                        ThrowHelper.ThrowInvalidOperationException_SerializerPropertyNameConflict(Type, property.Name);
+                    }
+
+                    isOrderSpecified = property.Order != 0;
+                }
+
+                if (isOrderSpecified)
+                {
+                    PropertyCache.List.StableSortByKey(static prop => prop.Value.Order);
+                }
+            }
+            else
+            {
+                // Resolver didn't modify properties
+
+                // Source gen currently when initializes properties
+                // also assigns JsonPropertyInfo's JsonTypeInfo which causes SO if there are any
+                // cycles in the object graph. For that reason properties cannot be added immediately.
+                // This is a no-op for ReflectionJsonTypeInfo
+                LateAddProperties();
+                PropertyCache ??= CreatePropertyCache(capacity: 0);
+            }
+
+            if (ExtensionDataProperty != null)
+            {
+                ExtensionDataProperty.EnsureChildOf(this);
+                ExtensionDataProperty.EnsureConfigured();
+            }
+
+            foreach (KeyValuePair<string, JsonPropertyInfo> jsonPropertyInfoKv in PropertyCache.List)
+            {
+                JsonPropertyInfo jsonPropertyInfo = jsonPropertyInfoKv.Value;
+
+                jsonPropertyInfo.EnsureChildOf(this);
+                jsonPropertyInfo.EnsureConfigured();
```

Comment:
why do we need to configure and set parent over here, shouldn't that happen directly in `Configure`? - also you can do it by iterating on `_properties` directly without having to special case extension data property

Reply:
basically it seems a bit weird that we're doing more than initializing cache in here

Reply:
This is logic being called by Configure, I've just factored it into a separate method.